File: question/leetcode/leetcode0025.Reverse-Nodes-in-k-Group/leetcode0025.go
package leetcode

import (
	"clarencedu/algorithm-go/structure"
)

type ListNode = structure.ListNode

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}
	// a,b代表k个节点的前后节点
	// 定位当前组的起始(a)和下一组的起始(b)
	a, b := head, head
	for i := 0; i < k; i++ {
		if b == nil {
			return head
		}
		b = b.Next
	}
	newHead := reverse(a, b)
	a.Next = reverseKGroup(b, k)
	return newHead
}

// 反转[a,b)区间的链表
func reverse(head *ListNode, target *ListNode) *ListNode {
	var pre, cur, next *ListNode
	pre, cur, next = nil, head, head
	for cur != target {
		next = cur.Next
		cur.Next = pre
		pre = cur
		cur = next
	}
	return pre
}

File: question/leetcode/leetcode0025.Reverse-Nodes-in-k-Group/leetcode0025.ts
class ListNode {
	val: number
	next: ListNode | null
	constructor(val?: number, next?: ListNode | null) {
		this.val = (val === undefined ? 0 : val)
		this.next = (next === undefined ? null : next)
	}

}

function reverseKGroupII(head: ListNode | null, k: number): ListNode | null {
	if (!head) {
		return null
	}
	let a: ListNode | null = head, b: ListNode | null = head
	for (let i = 0; i < k; i++) {
		if (!b) {
			return head
		}
		b = b.next

	}
	let newHead = reverseII(a, b)
	a.next = reverseKGroupII(b, k)
	return newHead;
};


function reverseII(head: ListNode | null, target: ListNode | null): ListNode | null {
	let pre: ListNode | null = null, cur: ListNode | null = head, next: ListNode | null = head;
	while (cur !== target) {
		if (!cur) {
			return pre
		}
		next = cur.next;
		cur.next = pre;
		pre = cur;
		cur = next;
	}
	return pre
}
